src/Features/Core/Portable/Workspace/MiscellaneousFileUtilities.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Scripting.Hosting;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Features.Workspaces
+{
+    internal static class MiscellaneousFileUtilities
+    {
+        internal static ProjectInfo CreateMiscellaneousProjectInfoForDocument(
+            string filePath,
+            TextLoader textLoader,
+            LanguageInformation languageInformation,
+            HostWorkspaceServices services,
+            ImmutableArray<MetadataReference> metadataReferences)
+        {
+            var fileExtension = PathUtilities.GetExtension(filePath);
+
+            var languageServices = services.GetLanguageServices(languageInformation.LanguageName);
+            var compilationOptions = languageServices.GetService<ICompilationFactoryService>()?.GetDefaultCompilationOptions();
+
+            // Use latest language version which is more permissive, as we cannot find out language version of the project which the file belongs to
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/575761
+            var parseOptions = languageServices.GetService<ISyntaxTreeFactoryService>()?.GetDefaultParseOptionsWithLatestLanguageVersion();
+
+            if (parseOptions != null &&
+                compilationOptions != null &&
+                fileExtension == languageInformation.ScriptExtension)
+            {
+                parseOptions = parseOptions.WithKind(SourceCodeKind.Script);
+                compilationOptions = GetCompilationOptionsWithScriptReferenceResolvers(services, compilationOptions, filePath);
+            }
+
+            var projectId = ProjectId.CreateNewId(debugName: "Miscellaneous Files Project for " + filePath);
+            var documentId = DocumentId.CreateNewId(projectId, debugName: filePath);
+
+            var sourceCodeKind = GetSourceCodeKind(parseOptions, fileExtension, languageInformation);
```

Comment:
Didn't we already figure out if this was a script a few lines above? I know this is a move but unsure if that cleans up some other things you're moving.